consensus/src/quorum_store/utils.rs

```diff
@@ -140,215 +141,274 @@ impl MempoolProxy {
             return_non_full = return_non_full,
             "Pull payloads from QuorumStore: internal"
         );
 
         if full || return_non_full {
-            counters::EXPIRED_PROOFS_WHEN_PULL.observe(num_expired_but_not_committed as f64);
             counters::BLOCK_SIZE_WHEN_PULL.observe(cur_txns as f64);
             counters::BLOCK_BYTES_WHEN_PULL.observe(cur_bytes as f64);
             counters::PROOF_SIZE_WHEN_PULL.observe(ret.len() as f64);
+            counters::EXCLUDED_TXNS_WHEN_PULL.observe(excluded_txns as f64);
             ret
         } else {
             Vec::new()
         }
     }
 
-    pub(crate) fn num_total_txns_and_proofs(&mut self, current_block_timestamp: u64) -> (u64, u64) {
-        let mut remaining_txns = 0;
-        let mut remaining_proofs = 0;
-        // TODO: if the digest_queue is large, this may be too inefficient
-        for (digest, expiration) in self.digest_queue.iter() {
-            // Not expired
-            if *expiration >= current_block_timestamp {
-                // Not committed
-                if let Some(Some(proof)) = self.digest_proof.get(digest) {
-                    remaining_txns += proof.num_txns();
-                    remaining_proofs += 1;
-                }
-            }
-        }
-        counters::NUM_TOTAL_TXNS_LEFT_ON_COMMIT.observe(remaining_txns as f64);
-        counters::NUM_TOTAL_PROOFS_LEFT_ON_COMMIT.observe(remaining_proofs as f64);
-
-        (remaining_txns, remaining_proofs)
-    }
+    pub(crate) fn handle_updated_block_timestamp(&mut self, block_timestamp: u64) {
+        assert!(
+            self.latest_block_timestamp <= block_timestamp,
+            "Decreasing block timestamp"
+        );
+        self.latest_block_timestamp = block_timestamp;
 
-    // returns the number of unexpired local proofs
-    pub(crate) fn clean_local_proofs(&mut self, current_block_timestamp: u64) -> Option<u64> {
-        let num_expired = self
-            .local_digest_queue
-            .iter()
-            .take_while(|(_, expiration_time)| *expiration_time < current_block_timestamp)
-            .count();
-        self.local_digest_queue.drain(0..num_expired);
-
-        let mut remaining_local_proof_size = 0;
-
-        for (digest, expiration) in self.local_digest_queue.iter() {
-            // Not expired. It is possible that the proof entry in digest_proof was already removed
-            // when draining the digest_queue but local_digest_queue is not drained yet.
-            if *expiration >= current_block_timestamp {
-                if let Some(entry) = self.digest_proof.get(digest) {
-                    // Not committed
-                    if entry.is_some() {
-                        remaining_local_proof_size += 1;
+        let expired = self.expirations.expire(block_timestamp);
+        let mut num_expired_but_not_committed = 0;
+        for key in &expired {
+            if let Some(mut queue) = self.author_to_batches.remove(&key.author()) {
```

Comment:
Do you have to remove it and then put it back? cannot you mutate the queue in place?

Reply:
When I try to mutate with `get_mut` the compiler complains that I have two mutable borrows of `self` along with the `self.dec_remaining()` below.